### PR TITLE
Adding some new functionality and documentation to DummyReader

### PR DESCRIPTION
**Pull request by jonoomph**
Adding the ability to add test frames `WriteFrame()`, with fake image and audio data. This will can be used in unit-tests, and will soon be used to verify some new audio improvements (coming soon).

### CODE REVIEW

**Comment by jonoomph** [comment]
@musteresel Please take a look and let me know what you think.

File: src/Frame.cpp [hunk 1]
```diff
@@ -480,6 +480,11 @@ const unsigned char* Frame::GetPixels()
 // Get pixel data (for only a single scan-line)
 const unsigned char* Frame::GetPixels(int row)
 {
+	// Check for blank image
```
Comment (jonoomph) [inline src/Frame.cpp:483]: I know... more magic stuff happening in the Frame class :grimacing:

File: tests/DummyReader_Tests.cpp [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+	CHECK_EQUAL(1, r.GetFrame(1)->GetPixels(700)[700] == 0); // black pixel
+	CHECK_EQUAL(1, r.GetFrame(1)->GetPixels(701)[701] == 0); // black pixel
+}
+
+TEST (DummyReader_Fake_Frame) {
+	// Create a default fraction (should be 1/1)
+	openshot::DummyReader r(openshot::Fraction(30, 1), 1920, 1080, 44100, 2, 30.0);
+	r.Open(); // Open the reader
+
+	// Let's create some test frames
```
Comment (jonoomph) [inline tests/DummyReader_Tests.cpp:72]: Creating some test Frame objects, with custom audio data (it doesn't really need to make sense), and then recalling those frames from DummyReader, and verifying our custom audio data is still intact.

**Comment by jonoomph** [comment]
This PR is loosely related to https://github.com/OpenShot/libopenshot/issues/447, and is related to some timeline audio merging improvements being worked on by @musteresel.

**Comment by ferdnyc** [comment]
> Adding the ability to add test frames `WriteFrame()`, with fake image and audio data.

I have to say, I _really_ **hate** that idea. Adding a write-API to the dummy _reader_? That's... I don't see the point of unit-testing something if it has to be fitted with special APIs that make no sense in the rest of the codebase. What is that even testing?

Let me propose a slightly less invasive approach. Instead of adding `WriteFrame()` to the DummyReader just so the tests can populate its cache, why not just give it a constructor that takes a `CacheMemory` object? The unit tests can create the cache, call `testcache.Add(frame)` to their heart's content, and then call `new openshot::DummyReader(testcache)` to wrap a Reader around the completed cache. That at least makes it work roughly the same way any other reader works, where it gets wrapped around some data (whether it's a file path, HTML source, etc...) in its constructor.

That way, the "special" API is limited to a constructor overload, and the rest of it is just a normal reader.

**Comment by jonoomph** [comment]
@ferdnyc :+1: I like that as well. Let me tweak things real quick.

File: src/Frame.cpp (hunk 1 above)
Comment (ferdnyc) [reply src/Frame.cpp:483]: Not just magic stuff, _wrong_ stuff. Getters shouldn't be creating data! Heck, that should be a `const` method.

If there's no image in the Frame and something requests pixel data, isn't the right thing to either throw an exception or return `nullptr`? Why does the class do so much work to make it **impossible** to keep image data out of it? I mean, heck, why is there even a `has_image_data` info member, when it's never checked, and it's virtually **impossible** for it to remain `false` even if you TRY?

I feel like all this magical image-creation is trying to make the API more "friendly" by loading it with assumptions about what the caller wants, which is only even maybe-right when they're **correct** assumptions. In fact, even the idea that auto-creating data out of thin air makes the API more friendly is _itself_ an assumption.

If nothing else the comment is wrong. The image will be filled with whatever the value of `color` is, not necessarily black. (And `color` will be set to that argument, at the same time. Which makes me really nervous, now that I look at it.)

**Comment by jonoomph** [comment]
We are going to be generating Frame objects with artificial audio data, such as an always incrementing number, and then build unittests which pass these frames through the Timeline with different FPS and sample rates. The idea is to verify audio data being truncated in the Timeline when combining Frames of different amounts of audio samples, and then correct the behavior once we can verify the failures. Hope that makes sense, haha.

**Comment by ferdnyc** [comment]
Like I said in the inline comments, this part suddenly makes me nervous:

https://github.com/OpenShot/libopenshot/blob/7831cfe9128502ae894d85c14e033991cd9a3d1b/src/Frame.cpp#L737-L756

If calling `GetPixelData()` can trigger an `AddColor()`, then theoretically it could be called from two threads at the same time. There's no reason to mistrust a getter that returns `const`, normally. But that `color = new_color;` bit is outside all of the thread synchronization. Yeah, it's going to be the _same_ string, but because we pass everything by value the code doesn't actually KNOW that, and it's going to copy it over again anyway, potentially colliding on the writes.

**Pull request merged**